js/noms/src/spec.js

```diff
@@ -0,0 +1,315 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// @flow
+
+import AbsolutePath from './absolute-path.js';
+import {invariant} from './assert.js';
+import {BatchStoreAdaptor} from './batch-store.js';
+import Dataset, {datasetRe} from './dataset.js';
+import Database from './database.js';
+import HttpBatchStore, {DEFAULT_MAX_READS} from './http-batch-store.js';
+import MemoryStore from './memory-store.js';
+import type Value from './value.js';
+
+export type SpecOptions = {
+  /**
+   * Authorization token for requests. For example, if the Database is backed by
+   * HTTP this will used for an `Authorization: Bearer ${authorization}` header.
+   */
+  authorization?: string;
+
+  /**
+   * If this is specified and > 0, the Database instances returned by
+   * `database()` will be backed by a cache.
+   */
+  cacheSize?: number;
+};
+
+/**
+ * Spec is a parsed specification for the location of a Noms database, dataset,
+ * or value.
+ *
+ * Spec has the same `Database` instance for its construction, and the
+ * accessors have the same mutability semantics as the `Database` interface,
+ * which is mutable. For example, if the underlying `Database` commits a
+ * dataset, then calls to `Spec.dataset()` will return the new one.
+ *
+ * For example:
+ *  - a database spec might be `mem` or `https://demo.noms.io/cli-tour`.
+ *  - a dataset spec might be `mem::photos` or `https://demo.noms.io/cli-tour::sf-crime`.
+ *  - a path spec might be `http://demo.noms.io::foo.bar` or `http://demo.noms.io::#abcdef.bar`
+ *
+ * See https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
+ */
+export default class Spec {
+  _spec: string;
+  _opts: ?SpecOptions;
+  _protocol: string;
+  _databaseName: string;
+
+  // Lazily created.
+  _database: ?Database;
+
+  // Exists if the spec was constructed through `forDataset()`.
+  _datasetName: ?string;
+
+  // Exists if the spec was constructed through `fromValue()` or `pin()`.
+  _path: ?AbsolutePath;
+
+  constructor(spec: string, databaseSpec: string, opts: ?SpecOptions) {
+    this._spec = spec;
+    this._opts = opts;
+    const [protocol, databaseName] = parseDatabaseSpec(databaseSpec);
+    this._protocol = protocol;
+    this._databaseName = databaseName;
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a database.
+   */
+  static forDatabase(spec: string, opts: ?SpecOptions): Spec {
+    return new Spec(spec, spec, opts);
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a dataset.
+   */
+  static forDataset(spec: string, opts: ?SpecOptions): Spec {
```

Comment:
I think you at least need to distinguish between a database and dataset/path. For example: IPV6 addresses contain `::` and you can't tell if that's a database separator or just part of the database.

But we could combine forDataset/forPath for parsing sake. The use cases are typically quite different, though: you want a dataset if you want to commit to a dataset, for example. You want a path if you want to extract a single value.

If they were combined, code would need to look like:

```javascript
const spec = Spec.parse('http://demo.noms.io/cli-tour::myDataset');
if (!spec.isDataset()) {
  console.error('must be a dataset');
  return;
}
```

The advantage of separate forDataset/forPath methods is that the sanity checking is done for you.